### PR TITLE
Add sysroot detection in rustdoc

### PR DESCRIPTION
**Pull request by robinmoussu**
When building `rustdoc` without using `x.py`, the sysroot is not going
to be detected automatically. It wasn't an issue when using `x.py`, and
this commit should not have any effect when building `rustdoc` as part
of the `rustc` build.

Note: if you want to build rustdoc out-of-tree, you need to install the
rustup components `rustc-dev` and `llvm-tools-preview`.

r? @simulacrum
@jyn514

### CODE REVIEW

File: src/librustdoc/lib.rs [hunk 1]
```diff
@@ -109,7 +142,25 @@ fn get_args() -> Option<Vec<String>> {
...
+            if let Some(sysroot) = compile_time_sysroot() {
+                let sysroot_flag = "--sysroot";
+                if !args.iter().any(|e| e == sysroot_flag) {
+                    // We need to overwrite the default that librustc_session would compute.
+                    args.push(sysroot_flag.to_owned());
+                    args.push(sysroot);
+                }
+            }
+
+            args
```
Comment (robinmoussu) [inline src/librustdoc/lib.rs:162]: I'm not 100% sure that it's the most readable way to update the `Vec<String>` in this `Option`.

File: src/librustdoc/lib.rs [hunk 2]
```diff
@@ -96,6 +96,39 @@ pub fn main() {
...
+    // For builds outside rustc, we need to ensure that we got a sysroot
+    // that gets used as a default.  The sysroot computation in librustc_session would
+    // end up somewhere in the build dir (see `get_or_default_sysroot`).
+    // Taken from PR <https://github.com/Manishearth/rust-clippy/pull/911>.
+    let home = option_env!("RUSTUP_HOME").or(option_env!("MULTIRUST_HOME"));
+    let toolchain = option_env!("RUSTUP_TOOLCHAIN").or(option_env!("MULTIRUST_TOOLCHAIN"));
+    Some(match (home, toolchain) {
+        (Some(home), Some(toolchain)) => format!("{}/toolchains/{}", home, toolchain),
+        _ => option_env!("RUST_SYSROOT")
+            .expect("To build Miri without rustup, set the `RUST_SYSROOT` env var at build time")
```
Comment (RalfJung) [inline src/librustdoc/lib.rs:127]: "Miri" here is probably wrong.

Comment (robinmoussu) [reply src/librustdoc/lib.rs:127]: Fixed. Is there anything else to change?

Comment (jyn514) [reply src/librustdoc/lib.rs]: This is waiting on review from `@Mark-Simulacrum` I think.

**Comment by Mark-Simulacrum** [comment]
r? @jyn514 

I think this is probably "fine" though I'm not personally really all that happy to see it. It seems like it may be worthwhile to consider putting this logic in rustc_driver (or similar), though, to avoid duplicating it into all the tools.

File: src/librustdoc/lib.rs [hunk 3]
```diff
@@ -96,6 +96,39 @@ pub fn main() {
...
+    if option_env!("RUSTC_STAGE").is_some() {
+        // This is being built as part of rustc, and gets shipped with rustup.
+        // We can rely on the sysroot computation in librustc_session.
+        return None;
+    }
+    // For builds outside rustc, we need to ensure that we got a sysroot
+    // that gets used as a default.  The sysroot computation in librustc_session would
+    // end up somewhere in the build dir (see `get_or_default_sysroot`).
+    // Taken from PR <https://github.com/Manishearth/rust-clippy/pull/911>.
+    if option_env!("RUSTC_STAGE").is_some() {
```
Comment (RalfJung) [inline src/librustdoc/lib.rs:113]: Why is this conditional duplicated?

**Comment by RalfJung** [comment]
At least for Miri, I don't see how to easily move this to rustc_driver. We specifically want to capture environment variables set during the Miri build, not when the librustc_driver for the sysroot gets built.

What would help, however, is if whoever instantiates the driver could affect `get_or_default_sysroot` to use a different default sysroot.

**Comment by Mark-Simulacrum** [comment]
Well, we could macro-ize it or something to get the expansion to be in the miri build I guess.

I imagine it should be possible to provide a global static of some kind that drivers can set to affect get_or_default_sysroot though.

**Comment by robinmoussu** [comment]
Where should this code be moved? In a macro in rustc_driver?

**Comment by RalfJung** [comment]
This code is a hack needed to avoid changing the code in rustc itself. IMO it makes no sense to move it to move it to the rustc repo. If rustc wants to support this usecase, instead `get_or_default_sysroot` should be changed to allow drivers to overwrite the default sysroot.

**Comment by RalfJung** [comment]
Specifically, this code hard-codes the default sysroot to be "one path above the binary":

https://github.com/rust-lang/rust/blob/a3d7a5e94507e3d6618790187bd85250ced82c4f/compiler/rustc_session/src/filesearch.rs#L128-L131

That doesn't work for Miri, which is why it needs that hack which this PR copies to rustdoc.

**Comment by flip1995** [comment]
FYI: Clippy does this too: https://github.com/rust-lang/rust-clippy/blob/408b615d3451d00c21b303fd8aeb74b77970a4a9/src/driver.rs#L301-L344

Summary:

```rust
        // Get the sysroot, looking from most specific to this invocation to the least:
        // - command line
        // - runtime environment
        //    - SYSROOT
        //    - RUSTUP_HOME, MULTIRUST_HOME, RUSTUP_TOOLCHAIN, MULTIRUST_TOOLCHAIN
        // - sysroot from rustc in the path
        // - compile-time environment
        //    - SYSROOT
        //    - RUSTUP_HOME, MULTIRUST_HOME, RUSTUP_TOOLCHAIN, MULTIRUST_TOOLCHAIN
```

The compile time sysroot is only the fallback for us. This is because we need `clippy_driver` to also work on single files, just like `rustc`, IIRC. (The discussion about this starts here: https://github.com/rust-lang/rust-clippy/pull/5071#issuecomment-577188796)

**Comment by robinmoussu** [comment]
I am not sure that I understand why I need to do to make this issue progress forward. Could you guide me?

**Comment by jyn514** [comment]
@robinmoussu right now the default sysroot is hard-coded in rustc. It should be configurable by tools (or any program using rustc_private) so they don't need to add the hack you're adding here. https://github.com/rust-lang/rust/pull/78926#issuecomment-726653035 has some suggestions on what code to change.

**Comment by crlf0710** [comment]
@robinmoussu Ping from triage, any updates on this?

**Comment by jyn514** [comment]
@crlf0710 this is a good refactor to have, but it's not really what @robinmoussu signed up for when he made the PR. Also it will probably not be used by rustdoc even if merged, since https://github.com/rust-lang/rust/pull/79540 is more the approach I see the team taking, we are not currently planning to switch rustdoc to building out of tree.

So if @robinmoussu wants to keep working on this more power to him, but I think it makes more sense for @RalfJung or @flip1995 to pick it up since they're the ones that will use it (or we can just make this an issue since the PR implementation is not on the right track).

**Comment by robinmoussu** [comment]
I agree with what @jyn514  said. Given that my bandwidth is more limited that what I'd like, I will not be able to work on it more than that. You can totally close this PR if you think it's not useful as-it.

**Comment by RalfJung** [comment]
Okay, let's close it then. Thanks @robinmoussu for exploring this solution, even if it did not work out!

**Pull request closed without merging**